services/workflows-service/src/collection-flow/collection-flow-state.service.ts

```diff
@@ -0,0 +1,235 @@
+      projectIds,
+    );
+
+    const entities = this.getEntityIdsFromWorkflow(workflowRuntimeData);
+
+    const documents = (await this.documentService.getByEntityIdsAndWorkflowId(
+      entities.map(entity => entity.entityId),
+      workflowId,
+      projectIds!,
+    )) as Document[];
+
+    const collectionFlowState = getCollectionFlowState(workflowRuntimeData.context);
+
+    if (!collectionFlowState) {
+      throw new NotFoundException('Collection flow state not found');
+    }
+
+    return this.computeCollectionFlowState(
+      uiDefinition,
+      workflowRuntimeData.context,
+      documents,
+      entities,
+    );
+  }
+
+  async updateCollectionFlowState() {}
+
+  private async computeCollectionFlowState(
+    uiDefinition: UiDefinition,
+    _context: AnyRecord,
+    documents: Document[],
+    entities: Array<{ entityId: string; entityType: 'ubo' | 'director' | 'business' }>,
+  ) {
+    const context = structuredClone(_context);
+    let documentsWithEntityTypes = documents.map(document => ({
+      ...document,
+      entityType:
+        entities.find(entity => entity.entityId === document.endUserId)?.entityType ||
+        entities.find(entity => entity.entityId === document.businessId)?.entityType,
+    }));
+
+    if (!getCollectionFlowState(context)) {
+      throw new NotFoundException('Collection flow state not found');
+    }
+
+    documentsWithEntityTypes = documentsWithEntityTypes.filter(
+      document =>
+        document.status === DocumentStatus.requested ||
+        document.decision === DocumentDecision.revisions,
+    );
+
+    setCollectionFlowStatus(context, CollectionFlowStatusesEnum.revision);
+
+    const collectionFlowSteps = (
+      uiDefinition.uiSchema as unknown as { elements: IUIDefinitionPage[] }
+    ).elements;
+
+    collectionFlowSteps.forEach(step => {
+      documentsWithEntityTypes.forEach(document => {
+        if (document.entityType === 'business') {
+          const businessDocumentDefinition = findBusinessDocumentDefinitionByTypeAndCategory(
+            step.elements,
+            {
+              type: document.type,
+              category: document.category,
+            },
+          );
+
+          if (businessDocumentDefinition) {
+            updateCollectionFlowStep(context, step.stateName, {
+              state: CollectionFlowStepStatesEnum.revision,
+            });
+          }
+        } else {
+          const entityFieldsDefinition = findEntityFieldsDefinition(
+            step.elements,
+            document.entityType!,
+          );
+
+          if (
+            entityFieldsDefinition &&
+            findDocumentDefinitionByTypeAndCategory(entityFieldsDefinition.children || [], {
+              type: document.type,
+              category: document.category,
+            })
+          ) {
+            updateCollectionFlowStep(context, step.stateName, {
+              state: CollectionFlowStepStatesEnum.revision,
+            });
+          }
+        }
```

Comment:
I think it's better here to throw an error if entityType is not defined or from our closed list. In the past we've made individual/business checks using ternary and else and it caused some bugs.